src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -588,19 +607,138 @@ public async Task<ImmutableArray<Diagnostic>> GetDiagnosticsAsync(Compilation co
 
-        public ImmutableArray<Diagnostic> DequeueLocalDiagnostics(DiagnosticAnalyzer analyzer, bool syntax, Compilation compilation)
+        public void ApplyProgrammaticSuppressions(DiagnosticBag reportedDiagnostics, Compilation compilation)
+        {
+            Debug.Assert(!reportedDiagnostics.IsEmptyWithoutResolution);
+            if (!_hasDiagnosticSuppressors)
+            {
+                return;
+            }
+
+            var newDiagnostics = ApplyProgrammaticSuppressionsCore(reportedDiagnostics.ToReadOnly(), compilation);
+            reportedDiagnostics.Clear();
+            reportedDiagnostics.AddRange(newDiagnostics);
+        }
+
+        public ImmutableArray<Diagnostic> ApplyProgrammaticSuppressions(ImmutableArray<Diagnostic> reportedDiagnostics, Compilation compilation)
+        {
+            if (reportedDiagnostics.IsEmpty ||
+                !_hasDiagnosticSuppressors)
+            {
+                return reportedDiagnostics;
+            }
+
+            return ApplyProgrammaticSuppressionsCore(reportedDiagnostics, compilation);
+        }
+
+        private ImmutableArray<Diagnostic> ApplyProgrammaticSuppressionsCore(ImmutableArray<Diagnostic> reportedDiagnostics, Compilation compilation)
+        {
+            Debug.Assert(_hasDiagnosticSuppressors);
+            Debug.Assert(!reportedDiagnostics.IsEmpty);
+            Debug.Assert(_programmaticSuppressions != null);
+
+            // We do not allow analyzer based suppressions for following category of diagnostics:
+            //  1. Diagnostics which are already suppressed in source via pragma/suppress message attribute.
+            //  2. Diagnostics explicitly tagged as not configurable by analyzer authors - this includes compiler error diagnostics.
+            //  3. Diagnostics which are marked as error by default by diagnostic authors.
+            var suppressableDiagnostics = reportedDiagnostics.WhereAsArray(d => !d.IsSuppressed &&
+                                                                                !d.IsNotConfigurable() &&
+                                                                                d.DefaultSeverity != DiagnosticSeverity.Error);
+            if (suppressableDiagnostics.IsEmpty)
+            {
+                return reportedDiagnostics;
+            }
+
+            ExecuteSuppressionActions(suppressableDiagnostics, concurrent: compilation.Options.ConcurrentBuild);
+            if (_programmaticSuppressions.IsEmpty)
+            {
+                return reportedDiagnostics;
+            }
+
+            var builder = ArrayBuilder<Diagnostic>.GetInstance(reportedDiagnostics.Length);
+
+            var programmaticSuppressionsByDiagnostic = _programmaticSuppressions.GroupBy(s => s.SuppressedDiagnostic)
+                .ToImmutableDictionary(g => g.Key,
+                                       g => new ProgrammaticSuppressionInfo(g.Select(s => (s.Descriptor.Id, s.Descriptor.Justification)).ToImmutableSortedSet()));
```

Comment:
>g.Select(s => (s.Descriptor.Id, s.Descriptor.Justification) [](start = 76, length = 59)

Is it impossible to have duplicates in this list?

Reply:
It is unlikely, but not impossible to have 2 separate suppressor packages choose the same ID and justification message.

Reply:
> It is unlikely, but not impossible to have 2 separate suppressor packages choose the same ID and justification message.

Wouldn't we want to be able to discover this fact (2 separate suppressor packages choose the same ID) from the log?

---
In reply to: [292220215](https://github.com/dotnet/roslyn/pull/36067#discussion_r292220215) [](ancestors = 292220215)

Reply:
Possibly, but we would then probably need to log the name (or full type name?) of the DiagnosticSuppressor to enable users to distinguish between duplicate suppressions. Not sure if it is worth adding this additional information in an already verbose diagnostic message for a very improbable scenario. We can always change this in future if we do hit this case or get user feedback.